parts/linux/cloud-init/artifacts/cse_helpers.sh

```diff
@@ -802,6 +802,40 @@ verify_DNS_health(){
     echo "DNS health check passed"
 }
 
+resolve_packages_source_url() {
+    local retries=5
+    local wait_sleep=1
+
+    for i in $(seq 1 $retries); do
+      response_code=$(curl -s -o /dev/null -w "%{http_code}" --max-time 5 --noproxy "*" https://packages.aks.azure.com/acs-mirror/healthz)
+      if [ ${response_code} -eq 200 ]; then
+        PACKAGE_DOWNLOAD_BASE_URL="packages.aks.azure.com"
+        echo "Setting PACKAGE_DOWNLOAD_BASE_URL to $PACKAGE_DOWNLOAD_BASE_URL."
+        break
+      else
+        if [ $i -eq $retries ]; then
+          PACKAGE_DOWNLOAD_BASE_URL="acs-mirror.azureedge.net"
+          echo "Setting PACKAGE_DOWNLOAD_BASE_URL to $PACKAGE_DOWNLOAD_BASE_URL. Please check to ensure cluster firewall has packages.aks.azure.com on its allowlist"
+          break
+        else
+          sleep $wait_sleep
+        fi
+      fi
+    done
+}
+
+update_base_url() {
```

Comment:
It will be great if you can write the shellspec unit test for this function. You can find many examples in cse_helpers_spec.sh

Reply:
There are also steps in the root's readme showing how to run the test